File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        int free=0,hold=-prices[0];
        for(int i=1;i<n;i++){
            int temp = hold;
            hold = max(hold, free - prices[i]);
            free = max(free,hold+prices[i]-fee);
        }
        return free;
    }
};

File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        n = len(prices)
        hold, free = -prices[0], 0
        
        for i in range(1,n):
            hold = max(hold,free-prices[i])
            free = max(free,hold+prices[i]-fee)
        
        return free